armv7a9-zynq7000/syspage.c

```diff
@@ -0,0 +1,501 @@
+			return -1;
+	}
+
+	syspage_common.maps[mapID].map.start = (u32)start;
+	syspage_common.maps[mapID].map.end = (u32)end;
+	syspage_common.maps[mapID].map.attr = attr;
+	syspage_common.maps[mapID].map.id = mapID;
+
+	size = plostd_strlen(name) < 7 ? plostd_strlen(name) : 7;
+	low_memcpy(syspage_common.maps[mapID].map.name, name, size);
+	syspage_common.maps[mapID].map.name[size] = '\0';
+
+	syspage_common.maps[mapID].top = (u32)start;
+
+	for (i = 0; i < MAX_ENTRIES_NB; ++i) {
+		if (syspage_common.entries[i].end != 0)
+			syspage_addEntries2Map(mapID, syspage_common.entries[i].start, syspage_common.entries[i].end);
+	}
+
+	syspage_common.mapsCnt++;
+
+	return 0;
+}
+
+
+int syspage_getMapTop(const char *map, void **addr)
+{
+	u8 id;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	*addr = (void *)syspage_common.maps[id].top;
+
+	return 0;
+}
+
+
+int syspage_alignMapTop(const char *map)
+{
+	u8 id;
+	u32 newTop;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	newTop = (syspage_common.maps[id].top + PAGE_SIZE - 1) & ~(PAGE_SIZE - 1);
+
+	if (newTop > syspage_common.maps[id].map.end)
+		return -1;
+
+	syspage_common.maps[id].top = newTop;
+
+	return 0;
+}
+
+
+int syspage_getFreeSize(const char *map, u32 *sz)
+{
+	u8 id;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	*sz = syspage_common.maps[id].map.end - syspage_common.maps[id].top;
+
+	return 0;
+}
+
+
+int syspage_write2Map(const char *map, const u8 *buff, u32 len)
+{
+	u8 id;
+	u32 freesz;
+
+	if (syspage_getMapID(map, &id) < 0) {
+		plostd_printf(ATTR_ERROR, "\nMAPS for %s doesn not exist!\n", map);
+		return -1;
+	}
+
+	freesz = syspage_common.maps[id].map.end - syspage_common.maps[id].top;
+
+	if (freesz < len ){
+		plostd_printf(ATTR_ERROR, "\nThere isn't any free space in %s !\n", map);
+		return -1;
+	}
+
+	if (syspage_isMapFree(id, len) < 0 ) {
+		plostd_printf(ATTR_ERROR, "\nUps!! You encountered on some data. Top has been moved. Please try again!!\n");
+		return -1;
+	}
+
+	low_memcpy((void *)syspage_common.maps[id].top, buff, len);
+
+	syspage_common.maps[id].top += len;
+
+	return 0;
+}
+
+
+void syspage_addEntries(u32 start, u32 sz)
+{
+	int i;
+
+	for (i = 0; i < MAX_ENTRIES_NB; ++i) {
+		if (syspage_common.entries[i].start == 0 && syspage_common.entries[i].end == 0) {
+			syspage_common.entries[i].start = start;
+			syspage_common.entries[i].end = start + sz;
+			break;
+		}
+	}
+
+	/* Check whether entry overlapped with existing maps */
+	for (i = 0; i < syspage_common.mapsCnt; ++i)
+		syspage_addEntries2Map(i, start, start + sz);
+}
+
+
+
+/* Program's functions */
+
+int syspage_addProg(void *start, void *end, const char *imap, const char *dmap, const char *name)
+{
+	u8 imapID, dmapID;
+	unsigned int pos = 0, len = 0;
```

Comment:
```suggestion
	unsigned int pos = 0, len;
```
If you accept following changes.